features/onboarding/Privy/usePrivySmartWalletConnection.ts

```diff
@@ -0,0 +1,116 @@
+import { useEmbeddedWallet, usePrivy } from "@privy-io/expo";
+import { usePrivyAccessToken, usePrivySigner } from "@utils/evm/privy";
+import { useEffect, useRef } from "react";
+
+import { usePrivyAuthStoreContext } from "./privyAuthStore";
+import { initXmtpClientFromViemAccount } from "@components/Onboarding/init-xmtp-client";
+import { ensureError, ensureErrorHandler } from "@utils/error";
+import logger from "@utils/logger";
+import { useSmartWallets } from "@privy-io/expo/smart-wallets";
+import { ViemAccount } from "@/utils/xmtpRN/signer";
+import config from "@/config";
+import { base } from "viem/chains";
+
+export function usePrivySmartWalletConnection(args: {
+  onConnectionDone: () => void;
+  onConnectionError: (error: Error) => void;
+  onStatusChange: (status: string) => void;
+}) {
+  const { onConnectionDone, onConnectionError, onStatusChange } = args;
+  const { isReady: privyIsReady, user: privyUser, logout } = usePrivy();
+  const embeddedWallet = useEmbeddedWallet();
+  const { client: smartWalletClient } = useSmartWallets();
+  const privyAccountId = usePrivyAuthStoreContext(
+    (state) => state.privyAccountId
+  );
+  usePrivySigner(true);
+  const privyAccessToken = usePrivyAccessToken();
+  const creatingEmbeddedWallet = useRef(false);
+  const initializingXmtp = useRef(false);
+
+  // Let's make sure we start with a clean state
+  useEffect(() => {
```

Comment:
We could clean up a lot of this state if we utilize React Query here. It would be a little odd in the sense that we would have to pass the client from usePrivy to the hook, since Privy doesn't look like it exposes any imperative APIs. 

But it would take care of cleaning up a lot of the management around whether or not we're currently creating and we would get the status changes for free and catching errors would be taken care of for free. Then, on success, we could initialize XMTP.

I'm happy to take a look at this with you to clean things up once you've got a final iteration on Privy and we've consolidated and agreed upon using it, which it seems like we already have.

Reply:
Another spot this would clean up: https://github.com/ephemeraHQ/converse-app/pull/1511#discussion_r1921617399